[bijection-util] Adding bijection for Java CompletableFuture

jessefeinman opened the pull request:
Problem: The current implementation for converting Java Futures to Twitter Futures is resource-intensive and doesn't special-case Java CompletableFutures which don't need timers or threadpools like a plain Java Future does

Solution: Java Completable Futures are a special case of Java Futures which provide an easy way to do a callback to complete a Twitter Future without the overhead of an extra thread pool or timer

Result: Java CompletableFutures can be easily be converted to and from Twitter Futures

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/bijection?pullRequest=425) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/bijection?pullRequest=425) before we can accept your contribution.<br/><hr/>**Jesse Feinman** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/bijection?pullRequest=425) it.</sub>

johnynek commented on bijection-util/src/main/scala/com/twitter/bijection/twitter_util/UtilBijections.scala:
```diff
@@ -133,6 +134,35 @@ trait UtilBijections {
+      override def apply(f: TwitterFuture[A]): CompletableFuture[A] = {
+        val cf = new CompletableFuture[A]()
+        f.respond {
+          case Return(value)    => cf complete value
```
minor nit, can we avoid infix (which is going away in scala 3). `cf.complete(value)`

jessefeinman replied:
No problem, do you want me to update the other places in this file where infix is being used? I mainly used it here due to its use elsewhere in the file.

johnynek replied:
I understand.

No need to change the previous code. We can run a scalafix or scalafmt or something to get other use cases, I just don't want to add new ones. I don't want to complicate the PR with other changes, but also don't want to dig in deeper on the old pattern.

Commit:
[bijection-util] Adding bijection for Java CompletableFuture

Problem: The current implementation for converting Java Futures to Twitter Futures is resource intensive and doesn't special case Java CompletableFutures
Solution: Java Completable Futures are a special case of Java Futures which provide an easy way to do a callback to complete the Twitter Future without the overhead of an extra threadpool or timer
Result: Java CompletableFutures can be easily be converted to and from Twitter Futures

mosesn commented on bijection-util/src/main/scala/com/twitter/bijection/twitter_util/UtilBijections.scala:
```diff
@@ -133,6 +134,35 @@ trait UtilBijections {
+  implicit def twitter2CompletableFutureBijection[A]
+      : Bijection[TwitterFuture[A], CompletableFuture[A]] = {
+    new AbstractBijection[TwitterFuture[A], CompletableFuture[A]] {
+      override def apply(f: TwitterFuture[A]): CompletableFuture[A] = {
```
nit: we have a slightly optimized CompletableFuture impl in Future

https://github.com/twitter/util/blob/develop/util-core/src/main/scala/com/twitter/util/Future.scala#L2159

so we can simplify this to `f.toCompletableFuture`

mosesn commented on bijection-util/src/main/scala/com/twitter/bijection/twitter_util/UtilBijections.scala:
```diff
@@ -133,6 +134,35 @@ trait UtilBijections {
+      }
+
+      override def invert(f: CompletableFuture[A]): TwitterFuture[A] = {
+        val p = new TwitterPromise[A]()
```
it would be cool to handle interruptions?  not a huge deal.

mosesn reviewed (commented):
sorry, this review was a bit late, figured I'd add some suggestions if we end up wanting to revisit this.
